exploration-module/src/main/java/io/edpn/backend/exploration/adapter/kafka/KafkaTopicHandler.java

```diff
@@ -0,0 +1,46 @@
+package io.edpn.backend.exploration.adapter.kafka;
+
+import io.edpn.backend.exploration.application.port.outgoing.CreateTopicPort;
+import lombok.RequiredArgsConstructor;
+import org.apache.kafka.clients.admin.AdminClient;
+import org.apache.kafka.clients.admin.ListTopicsResult;
+import org.apache.kafka.clients.admin.NewTopic;
+
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+@RequiredArgsConstructor
+public class KafkaTopicHandler implements CreateTopicPort {
+
+    private final AdminClient kafkaAdminClient;
+    private final int topicPartitions;
+    private final short topicReplicationFactor;
+
+
+    @Override
+    public CompletableFuture<Void> createTopicIfNotExists(String topicName) {
+        ListTopicsResult listTopicsResult = kafkaAdminClient.listTopics();
+        CompletableFuture<Void> topicCreationFuture = new CompletableFuture<>();
+
+        listTopicsResult.names().whenComplete((topicNames, exception) -> {
+            if (exception == null) {
+                if (!topicNames.contains(topicName)) {
+                    NewTopic newTopic = new NewTopic(topicName, topicPartitions, topicReplicationFactor);
+                    kafkaAdminClient.createTopics(Collections.singletonList(newTopic)).all().whenComplete((result, createException) -> {
+                        if (createException == null) {
+                            topicCreationFuture.complete(null);
+                        } else {
+                            topicCreationFuture.completeExceptionally(createException);
```

Comment:
Correct me if I'm wrong, but the logic here looks like check-then-act, having a (theoretical) window of opportunity between checking if a topic exists and creating it. If trying to create an existing topic raises an exception, I think in order to be 100% correct, we should check which exception `createException` was, and report success if the topic already existed? (at least that's how I interpret the "contract" implied by the method name)

Reply:
I just did a little digging on this.

Kafka documentation states that the method will "do nothing" if the topic already exists, which at least implies we should not see an exception (potentially, trying to create a thing that already exists, means a success).

However, you are indeed correct, that Kafka may take a number of seconds to create a topic. In a single-node world, I don't think we care. In a multi-node world, I think Kafka's "do nothing" principle means we don't need to care.

But perhaps it's a thing to test.